### PR TITLE
Expand searching of blockchain

### PR DESCRIPTION
**Pull request by SWvheerden**
## Description

Add search via kernel using the excess signature.
Adds search via STXO using the commitment.
Adds search via UTXO using the commitment.
Adds linking calls on the UI so users can use the CLI to call these calls.

## Motivation and Context

This allows users to find out if transactions was mined and or not. This will tell the user of the blockchain knows of Kernels, Utxos and or Stxos. If the BN is a pruned node, and the info if past pruning horizon, it will only tell the user that the Info exists, but it does not know which block. 

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/core/src/chain_storage/error.rs [hunk 1]
```diff
@@ -54,7 +51,7 @@ pub enum ChainStorageError {
     #[error("A parameter to the request was invalid")]
     InvalidQuery(String),
     #[error("The requested value '{0}' was not found in the database")]
-    ValueNotFound(DbKey),
+    ValueNotFound(String),
```
Comment (sdbondi) [inline base_layer/core/src/chain_storage/error.rs]: Why was this changed? The `{0}` calls DbKey::to_string() internally and the `ValueNotFound` error is only applicable for variants of `DbKey`.

File: base_layer/core/src/base_node/proto/request.rs [hunk 2]
```diff
@@ -86,6 +112,27 @@ impl From<ci::NodeCommsRequest> for ProtoNodeCommsRequest {
             FetchTxos(hash_outputs) => ProtoNodeCommsRequest::FetchTxos(hash_outputs.into()),
             FetchBlocks(block_heights) => ProtoNodeCommsRequest::FetchBlocks(block_heights.into()),
             FetchBlocksWithHashes(block_hashes) => ProtoNodeCommsRequest::FetchBlocksWithHashes(block_hashes.into()),
+            FetchBlocksWithKernels(signatures) => {
+                let mut sigs = Vec::new();
+                for sig in signatures {
+                    sigs.push(sig.into())
+                }
```
Comment (sdbondi) [inline base_layer/core/src/base_node/proto/request.rs]: nit:
```suggestion
                let sigs = signatures.into_iter().map(Into::into).collect();
```

Same for below. This will also only allocate `sigs` once (Vec::with_capacity)

File: base_layer/core/src/base_node/proto/request.rs [hunk 3]
```diff
@@ -50,6 +55,27 @@ impl TryInto<ci::NodeCommsRequest> for ProtoNodeCommsRequest {
             FetchTxos(hash_outputs) => ci::NodeCommsRequest::FetchTxos(hash_outputs.outputs),
             FetchBlocks(block_heights) => ci::NodeCommsRequest::FetchBlocks(block_heights.heights),
             FetchBlocksWithHashes(block_hashes) => ci::NodeCommsRequest::FetchBlocksWithHashes(block_hashes.outputs),
+            FetchBlocksWithKernels(signatures) => {
+                let mut sigs = Vec::new();
```
Comment (sdbondi) [inline base_layer/core/src/base_node/proto/request.rs:59]:
```suggestion
                let mut sigs = Vec::with_capacity(signatures.sigs.len());
```

Same for below

**Review by sdbondi — COMMENTED** [review]
LGTM - apart from the use of a string in the error type and some small optimisation nits.

File: base_layer/core/src/chain_storage/error.rs (hunk 1 above)
Comment (SWvheerden) [reply base_layer/core/src/chain_storage/error.rs]: Though this might come up. The problem is, I am not searching them via the usual keys. 
For example DB::key assumes you will always search a kernel via hash, we dont here, we search via the excess_sig. 

So if I cant find the error I have to create ValueNotFound(Kernel::hash), but I dont have the hash. So I cant create a value not found

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 4]
```diff
@@ -902,6 +922,105 @@ fn fetch_block<T: BlockchainBackend>(db: &T, height: u64) -> Result<HistoricalBl
...
+        for utxo in utxos.0 {
+            if utxo.commitment == commitment {
+                return Ok(None);
+            }
+        }
+        for comm in utxos.1 {
+            if comm == commitment {
+                return Ok(None);
+            }
+        }
```
Comment (sdbondi) [inline base_layer/core/src/chain_storage/blockchain_database.rs:998]: Destructuring the tuple is preferable

```suggestion
        let (outputs, spent) = fetch_outputs(db, utxo_hashes)?;
        for utxo in outputs {
            if utxo.commitment == commitment {
                return Ok(None);
            }
        }
        for commit in spent {
            if commit == commitment {
                return Ok(None);
            }
        }
```

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 5]
```diff
@@ -902,6 +922,105 @@ fn fetch_block<T: BlockchainBackend>(db: &T, height: u64) -> Result<HistoricalBl
...
+    }
+    let err = format!("Block with utxo {}", commitment.to_hex());
+    Err(ChainStorageError::ValueNotFound(err))
```
Comment (sdbondi) [inline base_layer/core/src/chain_storage/blockchain_database.rs]: If a value cannot be found and looking at the type signature of the function, I'd expect `Ok(None)` to be returned ?

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 6]
```diff
@@ -902,6 +922,105 @@ fn fetch_block<T: BlockchainBackend>(db: &T, height: u64) -> Result<HistoricalBl
...
+        let inputs = fetch_inputs(db, deleted)?;
+        for input in inputs {
+            if input.commitment == commitment {
+                return Ok(Some(fetch_block(db, i)?));
+            }
+        }
+    }
+    // data is not in the pruning horizon, we cannot check stxo's behind pruning horizon
+    let err = format!("Block with stxo {}", commitment.to_hex());
+    Err(ChainStorageError::ValueNotFound(err))
```
Comment (sdbondi) [inline base_layer/core/src/chain_storage/blockchain_database.rs]: `Ok(None)`?

**Review by sdbondi — COMMENTED** [review]
Ah I didn't see where you used the String error. However I don't think you need to return a ValueNotFound because `Ok(None)` already means that the value was not found.

**Comment by SWvheerden** [comment]
Ok(None) Means that the value was found, but we dont know which block it is in. 
So its a pruned mode, it can check that the utxo exists, but the utxo is older than the pruning horizon. Thats when Ok(None) is returned.

File: base_layer/core/src/chain_storage/blockchain_database.rs (hunk 5 above)
Comment (SWvheerden) [reply base_layer/core/src/chain_storage/blockchain_database.rs]: See my reply below

File: base_layer/core/src/chain_storage/blockchain_database.rs (hunk 6 above)
Comment (SWvheerden) [reply base_layer/core/src/chain_storage/blockchain_database.rs]: See my reply below

File: base_layer/core/src/base_node/comms_interface/inbound_handlers.rs [hunk 7]
```diff
@@ -252,6 +251,82 @@ where T: BlockchainBackend + 'static
                 }
                 Ok(NodeCommsResponse::HistoricalBlocks(blocks))
             },
+            NodeCommsRequest::FetchBlocksWithKernels(excess_sigs) => {
+                let mut blocks = Vec::<HistoricalBlock>::with_capacity(1);
```
Comment (stringhandler) [inline base_layer/core/src/base_node/comms_interface/inbound_handlers.rs]: Did you perhaps mean `with_capacity(excess_sigs.length())`?

File: base_layer/core/src/base_node/comms_interface/inbound_handlers.rs [hunk 8]
```diff
@@ -252,6 +251,82 @@ where T: BlockchainBackend + 'static
...
+                            "Could not provide requested block containing kernel with sig {} to peer because not \
+                             stored",
+                            sig.get_signature().to_hex(),
+                        ),
+                        Err(e) => warn!(
+                            target: LOG_TARGET,
+                            "Could not provide requested block containing kernel with sig {} to peer because: {}",
+                            sig.get_signature().to_hex(),
+                            e.to_string()
+                        ),
```
Comment (stringhandler) [inline base_layer/core/src/base_node/comms_interface/inbound_handlers.rs:275]: Should this error not get sent back to the client?

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 9]
```diff
@@ -902,6 +922,105 @@ fn fetch_block<T: BlockchainBackend>(db: &T, height: u64) -> Result<HistoricalBl
...
+                return Ok(Some(fetch_block(db, i)?));
+            }
+        }
+        for comm in utxos.1 {
+            if comm == commitment {
+                return Ok(Some(fetch_block(db, i)?));
+            }
+        }
+    }
+    // data is not in the pruning horizon, lets check behind that but only if if there is a pruning horizon
```
Comment (stringhandler) [inline base_layer/core/src/chain_storage/blockchain_database.rs]:
```suggestion
    // data is not in the pruning horizon, let's check behind that but only if there is a pruning horizon
```

File: base_layer/core/src/chain_storage/error.rs (hunk 1 above)
Comment (stringhandler) [reply base_layer/core/src/chain_storage/error.rs]: Having the DbKey gives two pieces of information 1. The entity we are searching for and 2. The value we are searching for. It actually also tells us that we are searching for it by the id. Maybe we should expand the ValueNotFound to take these params: Table/Entity name (string), Field being searched (string), Value (string), and then potentially the cause as an addition param?

File: base_layer/core/src/chain_storage/blockchain_database.rs (hunk 6 above)
Comment (stringhandler) [reply base_layer/core/src/chain_storage/blockchain_database.rs]: See my reply below as well, but wanted to highlight this occurrence. This code is assuming that the error message provided here will fit into the ValueNotFound formatted error's message. Splitting up the params guards against someone changing the formatted error message and making "Block with stxo..." look strange

**Review by stringhandler — COMMENTED** [review]
This will be really useful, thanks. I think we can discuss the ValueNotFound error a bit more and come to a neater solution, but after that it can be merged

File: base_layer/core/src/base_node/comms_interface/inbound_handlers.rs (hunk 7 above)
Comment (SWvheerden) [reply base_layer/core/src/base_node/comms_interface/inbound_handlers.rs]: Good catch, thats what it was, I played around with that code as I has other problems. Fixed this.

File: base_layer/core/src/base_node/comms_interface/inbound_handlers.rs (hunk 8 above)
Comment (SWvheerden) [reply base_layer/core/src/base_node/comms_interface/inbound_handlers.rs:275]: I kept it the same as all the other get_block function. 

They all return a vec<block>. This might only relate to a single block failing and the rest being fine.

File: base_layer/core/src/chain_storage/error.rs (hunk 1 above)
Comment (SWvheerden) [reply base_layer/core/src/chain_storage/error.rs]: I will create a struct for this all.

Comment (stringhandler) [reply base_layer/core/src/chain_storage/error.rs]: Why not individual params? Check out the InvalidHeader error here: https://docs.rs/thiserror/1.0.20/thiserror/

**Review by sdbondi — APPROVED** [review]
LGTM
